### PR TITLE
MAINT: Improve efficiency of pad by avoiding use of apply_along_axis

### PR DESCRIPTION
**Pull request by grlee77**
When the `mode` argument to `pad` is a function, `pad` assumes that apply_along_axis modifes the padded array in-place.

In this case, it is more efficient to avoid the overhead of creating an output buffer (as done internally within apply_along_axis) and just directly iterate the in-place operations using `transpose` and `nditer`.

This PR is an alternative to #13193. Aside from `pad`, I could not find any other instances in the NumPy or SciPy codebases where `apply_along_axis` was being used with functions operating `in-place`. In that case it may not be worth extending the API of `apply_along_axis`.

### CODE REVIEW

**Comment by charris** [comment]
close/reopen.

File: numpy/lib/arraypad.py [hunk 1]
```diff
@@ -797,9 +799,22 @@ def pad(array, pad_width, mode='constant', **kwargs):
...
-            np.apply_along_axis(
-                function, axis, padded, pad_width[axis], axis, kwargs)
+
+        in_dims = list(range(array.ndim))
+        for axis in in_dims:
+            # Iterate using ndindex as in apply_along_axis, but assuming that
+            # function operates inplace on the padded array.
+
+            # view with the iteration axis at the end
+            view = transpose(padded,
```
Comment (eric-wieser) [inline numpy/lib/arraypad.py]: Perhaps worth a comment with the equivalent `moveaxis` code (and a note that it would be slower)

**Review by eric-wieser — APPROVED** [review]
Looks sensible to me. I've seen some custom ndarray subclasses where transpose does not return a view, but probably fine to ignore those here.

File: numpy/lib/arraypad.py (hunk 1 above)
Comment (grlee77) [reply numpy/lib/arraypad.py]: `moveaxis` calls `transpose` internally in a similar fashion to what is proposed here.  `moveaxis` also returns a view, so performance should be very similar. I couldn't measure a consistent difference using `%timeit` on the docstring example and the difference should be even more negligible at larger sizes. This makes sense because `moveaxis` is only called once per axis while `function` is typically called many times per axis.

Given that
```Python
view = moveaxis(padded, axis, -1)
```

is easier to read than
```Python
view = transpose(padded,
                 in_dims[:axis] + in_dims[axis + 1:] + [axis])
```
I just switched to using `moveaxis` in the latest commit. Perhaps `transpose` was used internally for `apply_along_axis` mainly because that function predates the introduction of `moveaxis`?

Comment (eric-wieser) [reply numpy/lib/arraypad.py]: Thanks for measuring. My thinking was that since you already had the tuple of all axes constructed, and knew they were all in range, you could avoid `moveaxis` reconstructing the tuple and doing range checks each time. Sounds like the gain is so minimal that clarity is more important.

Comment (grlee77) [reply numpy/lib/arraypad.py]: Yes, I think we are on the same page. I initially implemented your suggested comment, but eventually decided it was clearer to just call `moveaxis`.

File: numpy/lib/arraypad.py [hunk 2]
```diff
@@ -797,9 +798,22 @@ def pad(array, pad_width, mode='constant', **kwargs):
         # Create a new zero padded array
         padded, _ = _pad_simple(array, pad_width, fill_value=0)
         # And apply along each axis
-        for axis in range(padded.ndim):
-            np.apply_along_axis(
-                function, axis, padded, pad_width[axis], axis, kwargs)
+
+        in_dims = list(range(array.ndim))
```
Comment (charris) [inline numpy/lib/arraypad.py]: Why is this needed? `axis in range` does the job.

Comment (eric-wieser) [reply numpy/lib/arraypad.py]: Leftover from before moveaxis was being used, good catch

Comment (grlee77) [reply numpy/lib/arraypad.py]: yes, thanks for catching that. I fixed it and then rebased this to a single commit.

**Comment by charris** [comment]
Thanks @grlee77 .
